imu_filter_madgwick/src/imu_filter_ros.cpp

```diff
@@ -21,149 +21,184 @@
+  }
+
+  // TODO: Add reconfigure parameter descriptions.
+  // gen.add("gain", double_t, 0, "Gain of the filter. Higher values lead to faster convergence but
+  // more noise. Lower values lead to slower convergence but smoother signal.", 0.1, 0.0, 1.0)
+  // gen.add("zeta", double_t, 0, "Gyro drift gain (approx. rad/s).", 0, -1.0, 1.0)
+  // gen.add("mag_bias_x", double_t, 0, "Magnetometer bias (hard iron correction), x component.", 0,
+  // -10.0, 10.0) gen.add("mag_bias_y", double_t, 0, "Magnetometer bias (hard iron correction), y
+  // component.", 0, -10.0, 10.0) gen.add("mag_bias_z", double_t, 0, "Magnetometer bias (hard iron
+  // correction), z component.", 0, -10.0, 10.0) gen.add("orientation_stddev", double_t, 0,"Standard
+  // deviation of the orientation estimate.", 0, 0, 1.0)
+
+  double gain;
+  declare_parameter("gain", 0.1);
+  get_parameter("gain", gain);
+  double zeta;
+  declare_parameter("zeta", 0.0);
+  get_parameter("zeta", zeta);
+  double mag_bias_x;
+  declare_parameter("mag_bias_x", 0.0);
+  get_parameter("mag_bias_x", mag_bias_x);
+  double mag_bias_y;
+  declare_parameter("mag_bias_y", 0.0);
+  get_parameter("mag_bias_y", mag_bias_y);
+  double mag_bias_z;
+  declare_parameter("mag_bias_z", 0.0);
+  get_parameter("mag_bias_z", mag_bias_z);
+  double orientation_stddev;
+  declare_parameter("orientation_stddev", 0.0);
+  get_parameter("orientation_stddev", orientation_stddev);
+  filter_.setAlgorithmGain(gain);
+  filter_.setDriftBiasGain(zeta);
+  RCLCPP_INFO(get_logger(), "Imu filter gain set to %f", gain);
+  RCLCPP_INFO(get_logger(), "Gyro drift bias set to %f", zeta);
+  mag_bias_.x = mag_bias_x;
+  mag_bias_.y = mag_bias_y;
+  mag_bias_.z = mag_bias_z;
+  orientation_variance_ = orientation_stddev * orientation_stddev;
+  RCLCPP_INFO(get_logger(), "Magnetometer bias values: %f %f %f", mag_bias_.x, mag_bias_.y,
+              mag_bias_.z);
 
   // **** register dynamic reconfigure
-  config_server_.reset(new FilterConfigServer(nh_private_));
-  FilterConfigServer::CallbackType f = boost::bind(&ImuFilterRos::reconfigCallback, this, _1, _2);
-  config_server_->setCallback(f);
+  parameters_client_ = std::make_shared<rclcpp::AsyncParametersClient>(
+      this->get_node_base_interface(), this->get_node_topics_interface(),
+      this->get_node_graph_interface(), this->get_node_services_interface());
 
-  // **** register publishers
-  imu_publisher_ = nh_.advertise<sensor_msgs::Imu>(
-    ros::names::resolve("imu") + "/data", 5);
+  parameter_event_sub_ = parameters_client_->on_parameter_event(
+      std::bind(&ImuFilterMadgwickRos::reconfigCallback, this, _1));
 
-  if (publish_debug_topics_)
-  {
-    rpy_filtered_debug_publisher_ = nh_.advertise<geometry_msgs::Vector3Stamped>(
-      ros::names::resolve("imu") + "/rpy/filtered", 5);
+  // **** register publishers
+  imu_publisher_ = create_publisher<sensor_msgs::msg::Imu>("~/imu/data", 5);
```

Comment:
This should be `"imu/data"`, if we want to keep the old topic names. Same for all other publishers and subscribers: remove `"~/"` from the topic name. The topics can (and should) still be remapped in the launch file, correct?

Reply:
Done. This places the topics in global namespace. But, yes, topics can still be remapped via launch file.